### PR TITLE
[autodiff] Add comment to {JVP,VJP}Cloner giving more context.

### PR DESCRIPTION
**Pull request by gottesmm**

### CODE REVIEW

**Comment by gottesmm** [comment]
@atrick what do you think. Can you understand the comment?

**Comment by gottesmm** [comment]
@jckarter @slavapestov would be interested as well in your thoughts.

**Comment by atrick** [comment]
Thanks for doing this. I just reviewed the comment without reading any background, and I've forgotten how this works... so I'm just assuming this is the right place for the discussion and there's nothing one needs to read first.

Long ago I remember seeing a forum post about this. It seems that you could just copy the relevant explanation/examples from that.

It helped a lot to read the comment backward. So an brief 1-2 line introduction is in order.

A small example with concrete types might help.

It's not clear how type U.TangentVector differs from the type U.

---
> `///  JVP(f)(t1)(t2) = (f(Ti), Df(Ti)(t2))`

It's somewhat confusing that the type `Ti` looks like is being applying to functions. Shouldn't it be this?

` ///  JVP(f)(t1)(t2) = (f(t1), Df(t2))`

??

---

For `VJP`: again, start of with a brief description. How does this different from JVP in layman's terms.

> `///  VJP(f)(t)(u) = (f(Ti), f*(Ti)(t2))`

Again I don't know what it means to apply a type to a function.

The notation is confusing. I think you can just say explicitly that `f*` is the pullback function.
The diagram is great, but in general it would help to just list the symbols that will be used in the formula next to their definition:
```
f: user function
Ti: vector type of user function input
t: input vector value
Df: derivitive
f*: pullback
```
and so on...

I'm guessing it should be this

` ///  VJP(f)(t)(u) = (f(t), f*(t)(t2))`

But where did `t2` come from?

The intermediate function of type `(U.TangentVector) -> (U, Ti.TangentVector)` isn't defined anywhere. Is this just conceptual? Do we generate a SIL function?

I'm confused at how the formulas above differ from the actual SIL design. I'd like to see a description of the actual SIL functions that are generated.

File: include/swift/SILOptimizer/Differentiation/JVPCloner.h [hunk 1]
```diff
@@ -29,7 +29,27 @@ namespace autodiff {
...
+///
+/// In words, the JVP(f)(t1) is a function that maps the vector Ti to a 2nd
+/// function that maps t2 in T to (f(t1), Df(t1)(t2)). The first element of this
+/// result tuple, is just the constant original result of f at the point t1, but
+/// the 2nd result is a linear approximation of f at t1 that varies linearly
+/// over the input value t2.
+///
+/// Operationally, this cloner takes in a function f and produces said tuple
+/// function, producing the parts of the jacobian needed to evaluate the full
+/// derivative.
```
Comment (dan-zheng) [inline include/swift/SILOptimizer/Differentiation/JVPCloner.h:52]: Personal preference: I would use simpler writing from the ["JVP and VJP functions" section](http://bit.ly/swift-autodiff-internals) of the Swift Differentiable Programming Implementation Overview, for readers less familiar with math:
* JVP/VJP are copies of the original function that also return a differential (linear approximation function) or a pullback (backpropagation function).
* Differential takes derivative values of specific input values (corresponding to differentiation parameter indices) and returns derivative values of specific output values (corresponding to differentiation result indices).
* Pullback takes gradient values of select output values and returns gradient values of specific input values.

I find this explanation to be a bit more math-lite and explanatory of the semantics ("derivative value of input" vs "input vector").

Comment (dan-zheng) [reply include/swift/SILOptimizer/Differentiation/JVPCloner.h:52]: We could also explain the transform operationally with a simple example. [From "Derivative code generation" section](https://docs.google.com/document/d/1_BirmTqdotglwNTOcYAW-ib6mx_jl-gH9Dbg4WmHZh0/edit#heading=h.7h3piv26ukj0):

```swift
// Original function.
@differentiable
func foo(_ x: Float) -> Float {
    return sin(x) * cos(x)
}

// Simplified SIL pseudocode.
sil @foo : $(Float) -> Float {
bb0(%x):
  %y1 = apply @sin(%x)
  %y2 = apply @cos(%x)
  %y3 = apply @mul(%y1, %y2)
  return %y3
}
```

```swift
// JVP: replaces all function applications with JVP applications.
sil @jvp_foo : $(Float) -> (Float, (Float) -> Float) {
bb0(%x):
  (%y1, %df_sin) = apply @jvp_sin(%x)
  (%y2, %df_cos) = apply @jvp_cos(%x)
  (%y3, %df_mul) = apply @jvp_mul(%y1, %y2)
  // Return tuple of original result and differential.
  return (%y3, { %dx in
    %dy1 = apply %df_sin(%dx)
    %dy2 = apply %df_cos(%dx)
    %dy3 = apply %df_mul(%dy1, %dy2)
    return %dy3
  })
}
```

```swift
// VJP: replaces all function applications with VJP applications.
sil @vjp_foo : $(Float) -> (Float, (Float) -> Float) {
bb0(%x):
  (%y1, %pb_sin) = apply @vjp_sin(%x)
  (%y2, %pb_cos) = apply @vjp_cos(%x)
  (%y3, %pb_mul) = apply @vjp_mul(%y1, %y2)
  // Return tuple of original result and pullback.
  return (%y3, { %dy3 in
    // All "adjoint values" in the pullback are zero-initialized.
    // %dx = 0, %dy1 = 0, %dy2 = 0
    (%dy1, %dy2) += %pb_mul(%dy3)
    (%dx) += %pb_cos(%dy2)
    (%dx) += %pb_sin(%dy1)
    return %dx
  })
}
```

Comment (dan-zheng) [reply include/swift/SILOptimizer/Differentiation/JVPCloner.h:52]: Any thoughts here? @rxwei @gottesmm @atrick

Comment (rxwei) [reply include/swift/SILOptimizer/Differentiation/JVPCloner.h:52]: I'd prefer a verbatim description combined with a SIL-specific example of the function types. In the change, I find the math notation a little confusing and disconnected from the notation used in the [proposal](https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md).
